### PR TITLE
 Os 181 Database composite index support 

### PR DESCRIPTION
**Pull request by pritha-tarento**
Added composite index support for SQLG, Neo4j.
Testing done for creation of index(postive case)

TEST CASES Indexing Scenario for Index:
On fresh db  for creating index(single(SI), composite(CI))
Case 1. [(A, B), C, D   ]                         CI(count: 1)  SI(Count: 2)
   On update  Added E -> (A,B,E)     CI(count: 2)   SI(count: 2)  ## new composite key gets created
              Added D -> (A,B,D)            CI(count: 1)   SI(count: 3)         ## D got ignored.
Case 2. [(A), C, D]                                CI(count: 0)  SI(count: 3) 
   On update Added E -> (A,E)         CI(count: 0)   SI(count: 4)      ## A,E composite key gets created 

On fresh db  for creating index(unique index (UI))
Case 1.  [A]                                            UI(count: 1)
   On update  Added B                       UI(count: 2)

### CODE REVIEW

File: java/registry/src/main/java/io/opensaber/registry/util/Indexer.java [hunk 1]
```diff
@@ -56,50 +67,75 @@ public void setUniqueIndexFields(List<String> indexUniqueFields) {
...
+     * @param label
+     * @param parentVertex
+     * @throws NoSuchElementException
+     */
     private void createNonUniqueIndex(Graph graph, String label, Vertex parentVertex) throws NoSuchElementException {
-
-        databaseProvider.createIndex(graph, label, indexFields);
-        updateIndices(parentVertex, indexFields, false);
+        databaseProvider.createIndex(graph, label, singleIndexFields);        
+        databaseProvider.createCompositeIndex(graph, label, compositeIndexFields);
```
Comment (indrajra) [inline java/registry/src/main/java/io/opensaber/registry/util/Indexer.java]: I think createCompositeIndex() function must be called from createIndex function above...where you call createNonUniqueIndex, createUniqueIndex ...

File: java/registry/src/main/java/io/opensaber/registry/sink/Neo4jGraphProvider.java [hunk 2]
```diff
@@ -103,13 +105,24 @@ public String getId(Edge edge) {
...
-        
+        }        
+    }
+    
+    @Override
+    public void createCompositeIndex(Graph graph, String label, List<String> propertyNames){
+        Neo4JGraph neo4jGraph = (Neo4JGraph) graph;
+        String properties = "";
+        for (String propertyName : propertyNames) {
+            properties = properties.isEmpty() ? propertyName : (properties + "," + propertyName);
```
[the remark this answers is not in the document]
Comment (pritha-tarento) [reply java/registry/src/main/java/io/opensaber/registry/sink/Neo4jGraphProvider.java]: Input list is formed linear comma seperated values. i am thinking i must have this.

File: java/registry/src/main/java/io/opensaber/registry/sink/SqlgProvider.java [hunk 3]
```diff
@@ -65,28 +65,65 @@ public void createIndex(Graph graph, String label, List<String> propertyNames) {
...
+            properties.add(vertexLabel.getProperty(propertyName).get());
+        }
+        ensureIndex(vertexLabel, IndexType.NON_UNIQUE, properties);
+    }
+    /**
+     * 
+     * @param label
+     * @return
+     */
+    private VertexLabel getVertexLabel(String label) {
```
[the remark this answers is not in the document]
Comment (pritha-tarento) [reply java/registry/src/main/java/io/opensaber/registry/sink/SqlgProvider.java]: getVertexLabel -> vertex. Done

File: java/registry/src/main/java/io/opensaber/registry/sink/SqlgProvider.java [hunk 4]
```diff
@@ -65,28 +65,65 @@ public void createIndex(Graph graph, String label, List<String> propertyNames) {
...
+     * @param propertyNames
+     */
+    private void createCompositeIndexByIndexType(Graph graph, String label, List<String> propertyNames) {
+        VertexLabel vertexLabel = getVertexLabel(label);
+        List<PropertyColumn> properties = new ArrayList<>();
 
+        for (String propertyName : propertyNames) {
+            properties.add(vertexLabel.getProperty(propertyName).get());
+        }
+        ensureIndex(vertexLabel, IndexType.NON_UNIQUE, properties);
```
[the remark this answers is not in the document]
Comment (pritha-tarento) [reply java/registry/src/main/java/io/opensaber/registry/sink/SqlgProvider.java]: Done.

File: java/registry/src/main/java/io/opensaber/registry/util/Indexer.java (hunk 1 above)
Comment (pritha-tarento) [reply java/registry/src/main/java/io/opensaber/registry/util/Indexer.java]: Done.

**Comment by pritha-tarento** [comment]
Neo4j Unique index I will add
https://neo4j.com/docs/cypher-manual/current/schema/constraints/#constraints-create-unique-constraint

**Pull request merged**